### PR TITLE
output/cloudv2: Compact histogram

### PR DESCRIPTION
**Pull request by codebien**
## What?

Implementation of a more compact Protobuf representation for the Histogram type. This new version only stores and pushes the histogram's significant buckets (non-zero).

## Why?

The current implementation only trims the zero at the limit of the histogram, doing a massive waste of memory if the histogram has a sparse representation. The new version instead allocates only for non-zero buckets.

## Related PR(s)/Issue(s)

Updates #3117

### CODE REVIEW

File: output/cloud/expv2/hdr.go [hunk 1]
```diff
@@ -37,14 +38,13 @@ const (
...
-	// in this way we can avoid extra allocation for not significant buckets.
-	// All the zero buckets in between are preserved.
-	Buckets []uint32
+	// It does not include counters for the untrackable values,
+	// because they contain exception cases and require to be tracked in a dedicated way.
+	Buckets map[uint32]uint32
+
+	// Indexes keeps an ordered slice of unique-seen buckets' indexes.
+	// It allows to iterate the buckets in order. It uses an ascendent order.
+	Indexes []uint32
```
Comment (codebien) [inline output/cloud/expv2/hdr.go:47]: We may consider using a Btree as an optimization. But I would like to achieve stability in the protocol before.

File: output/cloud/expv2/hdr.go [hunk 2]
```diff
@@ -104,92 +94,72 @@ func (h *histogram) addToBucket(v float64) {
...
-	if h.LastNotZeroBucket >= index {
-		panic("buckets is already bigger than requested index")
+// histogramAsProto converts the histogram into the equivalent Protobuf version.
+func histogramAsProto(h *histogram, time int64) *pbcloud.TrendHdrValue {
+	var (
+		counters []uint32
+		spans    []*pbcloud.BucketSpan
+	)
+
+	// allocate only if at least one item is available
```
[the remark this answers is not in the document]
Comment (mstoykov) [reply output/cloud/expv2/hdr.go:133]: As we only send observed metrics in the aggregation time period - it should not be possible. The first time we see a metric we create the histogram structure and add the first sample.

File: output/cloud/expv2/hdr.go [hunk 3]
```diff
@@ -104,92 +94,72 @@ func (h *histogram) addToBucket(v float64) {
...
-	case len(h.Buckets) == 0:
-		h.init(index)
-	case index < h.FirstNotZeroBucket:
-		h.prependBuckets(index)
-	case index > h.LastNotZeroBucket:
-		h.appendBuckets(index)
-	default:
-		h.Buckets[index-h.FirstNotZeroBucket]++
+	if h.Buckets == nil {
+		h.Buckets = make(map[uint32]uint32)
```
Comment (mstoykov) [inline output/cloud/expv2/hdr.go:103]: I think this should be done just when we make the histogram struct.

File: output/cloud/expv2/hdr.go [hunk 4]
```diff
@@ -104,92 +94,72 @@ func (h *histogram) addToBucket(v float64) {
...
-	// It requires a benchmark if it is better than just copy it.
-
-	newBuckets := make([]uint32, newLen)
-	copy(newBuckets[h.FirstNotZeroBucket-index:], h.Buckets)
-	h.Buckets = newBuckets
-
-	// Update the stats
-	h.Buckets[0] = 1
-	h.FirstNotZeroBucket = index
+	// insert in the middle
```
[the remark this answers is not in the document]
Comment (mstoykov) [reply output/cloud/expv2/hdr.go:120]: I am also for dropping the comment below

File: output/cloud/expv2/hdr.go [hunk 5]
```diff
@@ -104,92 +94,72 @@ func (h *histogram) addToBucket(v float64) {
...
-		// the slice capacity cap(a) rather than the length"
-		h.Buckets = h.Buckets[:newLen]
-	} else {
-		newBuckets := make([]uint32, newLen)
-		copy(newBuckets, h.Buckets)
-		h.Buckets = newBuckets
-	}
+		// if the current and the previous indexes are not consecutive
+		// consider as closed the current on-going span and start a new one.
+		if diff := h.Indexes[i] - h.Indexes[i-1]; diff > 1 {
```
Comment (mstoykov) [inline output/cloud/expv2/hdr.go:147]: For more optimal space use this likely should be 3+ as adding 1 span is at least 2 counters (offset and length) .

So we can add at least 2 zeros before we became even. "at least" here is as I don't know if there is some more overhead for this in the protobuf protocol

**Review by mstoykov — DISMISSED** [review]
LGTM! But we can probably fix some of the small fixes before merging.

The optimizations can likely wait a while and maybe benchmarks ;)

File: output/cloud/expv2/hdr.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (codebien) [reply output/cloud/expv2/hdr.go:133]: Without values not, but it is possible to have Indexes and Buckets empty when there are only _untrackable_ values that are tracked into the special extreme buckets.

File: output/cloud/expv2/hdr.go (hunk 3 above)
Comment (codebien) [reply output/cloud/expv2/hdr.go:103]: I move it to a constructor

File: output/cloud/expv2/sink_test.go [hunk 6]
```diff
@@ -16,7 +17,14 @@ func TestNewSink(t *testing.T) {
 		{metrics.Counter, &counter{}},
 		{metrics.Gauge, &gauge{}},
 		{metrics.Rate, &rate{}},
-		{metrics.Trend, &histogram{}},
+		{
+			metrics.Trend,
+			&histogram{
```
Comment (olegbespalov) [inline output/cloud/expv2/sink_test.go:22]: newHistogram maybe :sweat_smile:

**Review by olegbespalov — APPROVED** [review]

File: output/cloud/expv2/sink_test.go (hunk 6 above)
Comment (codebien) [reply output/cloud/expv2/sink_test.go:22]: I was a bit unsure about it, but thinking more about it, it makes sense.

**Review by olegbespalov — APPROVED** [review]
Great job! :rocket:

File: output/cloud/expv2/hdr.go (hunk 1 above)
Comment (mstoykov) [reply output/cloud/expv2/hdr.go:47]: I wonder if we can't just order the buckets indexes at the end :shrug:. There is no need to constantly keep track of it -if we only want it to be in order :shrug:

**Review by mstoykov — APPROVED** [review]

File: output/cloud/expv2/hdr.go (hunk 1 above)
Comment (codebien) [reply output/cloud/expv2/hdr.go]: Yeah, it simplifies the code a lot, good suggestion.

**Review by mstoykov — APPROVED** [review]

**Review by olegbespalov — APPROVED** [review]

**Pull request merged**